### ISSUE TITLE
k8sWatcher.GetCachedPod missing some Pods, causing all traffic to be dropped for said Pods

### ISSUE THREAD

**Issue by Hayajiro** [cilium/cilium#24391]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When running Cilium 1.13.0 on Kubernetes 1.24.10 (also happened on 1.25.6), we noticed intermittent connectivity issues for pods on one specific node.

At some point (it seems like this happens after the connection to `kube-apiserver` is interrupted for a short moment), Cilium-Agent on one node (the issue seems to affect only one node at a time) stops knowing about newly scheduled Pods.  
This results in all egress traffic for this pod to be dropped with the following messages in the policy verdict log:
```
Policy verdict log: flow 0x605bbdfd local EP ID 1764, remote ID 52735, proto 17, egress, action deny, match none, 10.244.10.43:37507 -> 10.244.9.109:53 udp
```
Additionally, `cilium monitor -t drop` shows the following:
```
xx drop (Policy denied) flow 0x3d0de7b5 to endpoint 0, ifindex 126, file 2:1166, , identity init->52735: 10.244.10.43:59599 -> 10.244.11.81:53 udp
```

In this specific example, the pod tries to do a DNS lookup, but its requests to both CoreDNS pods are dropped.

When this issue occurs, the following is spammed all the time in the logs with `debug: true`:
```
level=debug msg="stopped waiting for caches to be synced" kubernetesResource="core/v1::Pods" subsys=k8s
level=warning msg="Unable to fetch kubernetes labels" containerID= datapathPolicyRevision=15 desiredPolicyRevision=15 endpointID=1764 error="pod.core \"hello-1-27981521-vptb6\" not found" identity=5 ipv4= ipv6= k8sPodName=/ subsys=resolve-labels
level=debug msg="Controller func execution time: 253.017µs" name=resolve-labels-default/hello-1-27981521-vptb6 subsys=controller uuid=2c18677b-9f69-4145-86e3-26a8d9ed4b3d
level=debug msg="Controller run failed" consecutiveErrors=26 error="pod.core \"hello-1-27981521-vptb6\" not found" name=resolve-labels-default/hello-1-27981521-vptb6 subsys=controller uuid=2c18677b-9f69-4145-86e3-26a8d9ed4b3d
level=debug msg="stopped waiting for caches to be synced" kubernetesResource="core/v1::Pods" subsys=k8s
level=warning msg="Unable to fetch kubernetes labels" containerID= datapathPolicyRevision=15 desiredPolicyRevision=15 endpointID=1764 error="pod.core \"hello-1-27981521-vptb6\" not found" identity=5 ipv4= ipv6= k8sPodName=/ subsys=resolve-labels
level=debug msg="Controller func execution time: 242.82µs" name=resolve-labels-default/hello-1-27981521-vptb6 subsys=controller uuid=2c18677b-9f69-4145-86e3-26a8d9ed4b3d
level=debug msg="Controller run failed" consecutiveErrors=25 error="pod.core \"hello-1-27981521-vptb6\" not found" name=resolve-labels-default/hello-1-27981521-vptb6 subsys=controller uuid=2c18677b-9f69-4145-86e3-26a8d9ed4b3d
[...]
```
According to our OpenSearch instance, this all happened within the same millisecond.

We only noticed this issue since upgrading to Cilium 1.13 (following the [upgrade guide at docs.cilium.io](https://docs.cilium.io/en/v1.13/operations/upgrade/)). With Cilium 1.12, we haven't seen this issue so far, even in highly used environments. From my limited understanding of Golang, it seems like it's related to changes introduced in 1.13 regarding the k8s-client and/or -caching. 

### Cilium Version

```
Client: 1.13.0 c9723a8d 2023-02-15T14:18:31+01:00 go version go1.19.6 linux/amd64
Daemon: 1.13.0 c9723a8d 2023-02-15T14:18:31+01:00 go version go1.19.6 linux/amd64
```

### Kernel Version

```
Linux m1.k8s-test-1 5.15.0-67-generic #74-Ubuntu SMP Wed Feb 22 14:14:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

### Kubernetes Version

```
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.10", GitCommit:"5c1d2d4295f9b4eb12bfbf6429fdf989f2ca8a02", GitTreeState:"clean", BuildDate:"2023-01-18T19:15:31Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.10", GitCommit:"5c1d2d4295f9b4eb12bfbf6429fdf989f2ca8a02", GitTreeState:"clean", BuildDate:"2023-01-18T19:08:10Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

### Sysdump

Can be provided upon request.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Hayajiro** [comment]
We were able to reproduce this, in a completely fresh cluster, with Cilium 1.3.1, so it seems this version is also affected.

**Comment by julianwiedmann** [comment]
Hi, thanks for reporting this issue!

For initial triage:
1. is this a persistent problem, or does cilium eventually recover and is able to resolve the affected pod's labels?
2. have you tried reproducing with wireguard disabled, to isolate whether those packet errors are related?
3. could you attach a sysdump, to clarify your configuration?

> From my limited understanding of Golang, it seems like it's related to changes introduced in 1.13 regarding the k8s-client and/or -caching.

Not sure I follow - did you already identify a problematic code section?

**Comment by Hayajiro** [comment]
Hi Julian,

> is this a persistent problem, or does cilium eventually recover and is able to resolve the affected pod's labels?

Once the issue appears, cilium is unable to recover by itself. You have to restart the respective `cilium-agent` pod on the affected node to temporarily fix the issue.

> have you tried reproducing with wireguard disabled, to isolate whether those packet errors are related?

We haven't tried this yet (as we require pod to pod traffic to be encrypted in our setup), but could possibly give this a try in one of our test environments.

>  could you attach a sysdump, to clarify your configuration?

We have a sysdump of the most recent ready (it happened again since yesterday). We'd prefer to send it via a private channel (E-Mail, Slack) if possible, though.

> did you already identify a problematic code section?

No, I didn't find anything concrete yet (I'm sadly not that well-versed in Golang), just that it has likely to do with the k8s-client and the implemented cache (which is populated by the k8s watcher). Otherwise I probably would've provided a patch already. :slightly_smiling_face: 

If you need anything else, just ask and we'll provide it ASAP.

**Comment by timbuchwaldt** [comment]
Weighing in here, we see the issue on 1.12.8 as well. The cluster was previously on 1.12.4, not exhibiting this bug. Other clusters on 1.12.7 with the same config are also fine from what I can tell. I'm looking into this further and hope to provide a sysdump, but that needs some further checking with regards to the included information.

**Comment by ReillyBrogan** [comment]
> Weighing in here, we see the issue on 1.12.8 as well. The cluster was previously on 1.12.4, not exhibiting this bug. Other clusters on 1.12.7 with the same config are also fine from what I can tell. I'm looking into this further and hope to provide a sysdump, but that needs some further checking with regards to the included information.

If you're sure that the issue is present in 1.12.8 but not in 1.12.7 then you can `git bisect` to figure out the bad commit. Assuming you are running Cilium as a container image you should just need to clone the cilium repo and run `make docker-cilium-image` to build the image. Then you can push it to a registry of your choice and then switch your test cluster to using it.

**Comment by christarazi** [comment]
@timbuchwaldt Are you looking into the bisect route that @ReillyBrogan suggested?

**Comment by timbuchwaldt** [comment]
Earliest could be mid-end of next week as I am busy climbing up mountains until then :D

**Comment by Hayajiro** [comment]
I'm not quite sure if the issue Tim is experiencing is actually related to this one right here. I've been trying to reproduce this on the same cluster that was having the issues before, just with Cilium 1.12.8, but had no luck so far. With Cilium 1.13 on the other hand, it's been very easy to trigger the issue within a couple minutes.

All you need is a handful of cronjobs running at every minute. Then, drop connections to the API server for a second (e.g. by killing the respective LB) around the time the jobs would be scheduled. This reliably triggers the issue for us after a few tries.

**Comment by Hayajiro** [comment]
I feel like this might've been fixed by #24870 - been trying to reproduce the bug on 1.13.2 but haven't been able to immediately trigger it like I've been able to on earlier releases of 1.13. So perhaps this was actually related to the bug in the k8s api client.

I'll continue testing tomorrow to see if I can manage to reproduce it, but this looks promising so far.

**Comment by Hayajiro** [comment]
Just a heads-up: The issue seems to have resurfaced with Cilium 1.13.3, we are seeing the exact same issue in our staging environment.

Will dig a bit deeper in a bit, but if anyone else has an idea, I'll appreciate any input.
